### DISCUSSION TITLE
"Could not delete databased file" everytime I open the game

### DISCUSSION THREAD

**Discussion by rvnexe** [ppy/osu#31299] [Q&A] [answered]
### Type

Game behaviour

### Bug description

Hi. It's been a while since I get 4 "Could not delete databased file" errors everytime I launch the game, and now I'm tired.
![image](https://github.com/user-attachments/assets/379f69b5-7c91-46cf-8c3b-b628a69049b2)

One day, I was making skins and while I was importing a .osk file, my PC got a BSOD (a service failure as I remember), and since then I'm getting those errors everytime I launch. I'm so sorry that I didn't make an issue sooner... looks like the logs are removed since then.

(the runtime logs are like this since they contain a high amount of those errors)
![image](https://github.com/user-attachments/assets/559fe1a4-34bc-4d67-aa24-ffb5a9ac3a44)

### Version

2024.1224.1

### Logs

[compressed-logs.zip](https://github.com/user-attachments/files/18255179/compressed-logs.zip)
[1735242471.runtime.log](https://github.com/user-attachments/files/18255181/1735242471.runtime.log)

**Comment by cihe13375** [comment]
there are lots of entries like this in the log:
```
2024-12-26 19:47:52 [error]: Could not delete databased file d3e9701b1f24a85352e401225eb5eead5cab83c16b75cb4033354bfcbca97512
2024-12-26 19:47:52 [error]: System.UnauthorizedAccessException: Access to the path 'X:\Games\osu-lazer\files\d\d3\d3e9701b1f24a85352e401225eb5eead5cab83c16b75cb4033354bfcbca97512' is denied.
```
considering when the errors start to appear according your description my first guess is that the filesystem is corrupted. Try if the files can be opened directly, if not then consider using tools like chkdsk

the second guess would be file permission issues, check the permission settings and make sure the file can be edited and deleted without admin privileges

**Reply by bdach** [reply]
As above, seems like a you problem unfortunately.

If you attach `client.realm` I might be able to tell you what these broken files are linked to, maybe you can delete the relevant item from the database to fix it or something.

**Reply by rvnexe** [reply]
> Try if the files can be opened directly

Those files are stored by their hash, there's no way I can open them if I don't know (at least) their extension. (as stated [here](https://osu.ppy.sh/wiki/en/Client/Release_stream/Lazer/File_storage))
> my first guess is that the filesystem is corrupted.

The disk seems healthy. I can interact with the files. Copy, cut, even delete them manually.
![image](https://github.com/user-attachments/assets/f41aadad-cb87-44fe-89bc-e8d3d627b2d8)

> the second guess would be file permission issues

Also the Permissions are looking correct, at least for osu! to be able to delete them.
![image](https://github.com/user-attachments/assets/6c0630b3-74be-40b9-9424-dc51aea3c1a7)
Even though, I ran osu! elevated, but same thing happened.
![image](https://github.com/user-attachments/assets/9fa1108b-69da-4381-ab57-8fe8053e9a98)
I picked a random logged file, and modified its permissions to full control, then ran osu! again. (in this case, `X:\Games\osu-lazer\files\6\63\633532cf80465a18f173edfd624a3b2d41b8829fdd76de29da380b37c83a57fd`)
![image](https://github.com/user-attachments/assets/86ab9dfd-14b1-4c2f-96c5-219a61033a00)

and it was in the new log again:
![image](https://github.com/user-attachments/assets/ad7b7481-ae71-4c03-9246-ee162b75440e)


With my few coding experience, I guess it's just osu!'s problem?

**Reply by rvnexe** [reply]
(in case you need it for further investigations:)
[client.realm.zip](https://github.com/user-attachments/files/18259538/client.realm.zip)

**Reply by bdach** [reply]
> Even though, I ran osu! elevated, but same thing happened.

Never do this. There is no reason to do this.

I'll look at the realm db later, but do you maybe have some antivirus software or something that may be quarantining the files or doing something weird with them?

**Reply by rvnexe** [reply]
> I'll look at the realm db later, but do you maybe have some antivirus software or something that may be quarantining the files or doing something weird with them?

No antivirus. (unless if we count windows defender as one)

**Reply by bdach** [reply]
To noone's surprise the files do not appear to be referenced by anything anymore in the realm database.

A few other things to try:

- You could try putting one of the broken files here in a `.zip` for inspection
- You could also try and see if you can delete at least one of the broken files yourself. If you can't, maybe windows explorer will provide some sort of explanation why.

**Reply by rvnexe** [reply]
I can delete them manually, I've mentioned earlier.
I did one (`a289249c66c935c6ab81dcca894ad5839f827d1fc94fe2c338cdd832793a6bb4`)
And it disappeared from the log:
![image](https://github.com/user-attachments/assets/a72d649b-b6e9-40e6-9705-b11930122cb0)
[some-randomly-picked-ones.zip](https://github.com/user-attachments/files/18259632/some-randomly-picked-ones.zip)

**Reply by bdach** [reply]
Okay so first of all after a bunch of weird stuff I recovered this from one of the files you attached (namely `9df17ecef1158648078a015fb439f5c07036a2a00c046377750dcbff20230750`):

![aa](https://github.com/user-attachments/assets/b792eaab-ab71-4c88-a213-9befa0f98408)

I did this by zlib uncompressing the file after `file` hinted it may work:

```
9df17ecef1158648078a015fb439f5c07036a2a00c046377750dcbff20230750: zlib compressed data
```

after which I got

```
9df17ecef1158648078a015fb439f5c07036a2a00c046377750dcbff20230750: Git blob 23112
```

so I trimmed off a few bytes off the start and got that image back. Any idea what that may be about? Do you recognise the image at all?

That is thing 1; the other thing is that when I unzipped the zip you provided, all the files within were marked readonly for whatever reason:

```
 ╰─ ls -la
total 64
drwxrwxr-x  2 dachb dachb  4096 gru 27 11:36 .
drwxr-xr-x 14 dachb dachb 20480 gru 27 11:36 ..
-r--r--r--  1 dachb dachb 23014 lis  8  2023 9df17ecef1158648078a015fb439f5c07036a2a00c046377750dcbff20230750
-r--r--r--  1 dachb dachb  7408 lis  8  2023 b61d58e1777c93269ca4b849df743c75659f0c641bebd3861951ed4da313ae8f
-r--r--r--  1 dachb dachb  2730 lis  8  2023 f5d6fe592680c3c013f8d837f55dafa6c7b3a9e9dd24b3044d8bacbe9bed3b73
```

So I'm thinking this may be an issue here too. The easiest solution would be probably to recursively mark the `files` folder as NOT readonly via explorer.

**Reply by rvnexe** [reply]
Umm.. This is exactly a sprite of the Skin that I mentioned earlier: 

> while I was importing a .osk file, my PC got a BSOD (a service failure as I remember), and since then I'm getting those errors

And yeah, those files were Read-Only! (smh I didn't notice) I removed Read-Only from one of those, and it got removed by osu! client successfully.
but I haven't touched any of them. Neither I nor any other programs I guess. I'm guessing that osu! client was importing those from the skin file, sorting them in the database, and making their respective hash-based files, but Read-Only?! Maybe it was gonna put them out from Read-Only state once the import is done, but it got interrupted?! Am I right?

**Reply by rvnexe** [reply]
Though, I de-Read-Only-ied (tf is grammar) the whole "files" folder, and the issue is solved for now. Thanks for the whole effort you put into my problem <3

**Reply by bdach** [reply, marked as answer]
> And yeah, those files were Read-Only! (smh I didn't notice) I removed Read-Only from one of those, and it got removed by osu! client successfully.
> but I haven't touched any of them. Neither I nor any other programs I guess. I'm guessing that osu! client was importing those from the skin file, sorting them in the database, and making their respective hash-based files, but Read-Only?! Maybe it was gonna put them out from Read-Only state once the import is done, but it got interrupted?! Am I right?

My guess would be that either the bsod or the post-bsod chkdsk (or whatever it was that windows ran) marked the files as read-only. So this is basically in the "freak accident" category of events. I'm not aware of osu! itself marking any file as readonly.

Let's just mark this as resolved for now.

**Comment by Chiyonofujifangirl** [comment]
I've recently come across this same issue with Lazer ever since I installed my new CPU, and for some reason it completely corrupts some of my maps alongside this, and prevents me from reinstalling them manually through downloading on the website.

Is there anything I can feasibly do to reset some odd setting in windows to make this go away, because as it seems they're not even setup to be read-only, and need to be viewed in admin mode in order to be accessed.

**Reply by Chiyonofujifangirl** [reply]
I have gone through and attempted running the game in admin mode to see if it repairs it (which is probably a generally bad idea), but the fact remains that some of said maps are still corrupted by this weird database problem.

**Reply by peppy** [reply]
If you don't mind resetting everything, you can just delete `client.realm`. If you'd like to recover data, let me know and I can probably give you some guidance.

**Reply by Chiyonofujifangirl** [reply]
I do want to retain my maps and the hundreds of replays i do have set over my time playing, so I really don't think just _deleting_ `client.realm` would be an option.

I hope that there is some way to try and fix this issue without resorting to just deleting all of the data, even if its' meticulous on the user end.

**Reply by peppy** [reply]
Please share a screenshot of your osu folder to start with, so we can see what state you're in.

**Reply by Chiyonofujifangirl** [reply]
![image](https://github.com/user-attachments/assets/567466ac-2048-4481-9ca1-8a12b083fe75)
Honestly I don't even know what else to look for, unless you also want my osu stable folder.

**Reply by peppy** [reply]
It doesn't look like you have any backup / previous versions of `client.realm`, so there's likely not much that can be done about your situation unless you have a backup of the game somewhere else.

**Reply by Chiyonofujifangirl** [reply]
I do not have a backup of the game, but is there anything in the logs to be able to have a general idea of what's causing these issues in the first place?

Whenever I try to re-import the beatmaps that are corrupted, it fails every single time.
